train_scripts/train_sana_sprint_diffusers.py

```diff
@@ -0,0 +1,1823 @@
+                        z_new = torch.randn_like(model_input) * sigma_data
+                        x_t_new = torch.cos(t_new) * model_input + torch.sin(t_new) * z_new
+
+                        F_theta = transformer(
+                            hidden_states=(x_t_new / sigma_data),
+                            timestep=t_new.flatten(),
+                            encoder_hidden_states=prompt_embeds,
+                            encoder_attention_mask=prompt_attention_mask,
+                            guidance=(scm_cfg_scale * args.guidance_embeds_scale),
+                            return_logvar=False,
+                            jvp=False,
+                        )
+
+                        pred_x_0 = torch.cos(t_new) * x_t_new - torch.sin(t_new) * F_theta * sigma_data
+
+                    # Sample timesteps for discriminator
+                    timesteps_D = compute_density_for_timestep_sampling_scm(
+                        batch_size=bsz,
+                        logit_mean=args.logit_mean_discriminator,
+                        logit_std=args.logit_std_discriminator,
+                    )
+                    t_D = timesteps_D.view(-1, 1, 1, 1)
+
+                    # Add noise to predicted x0
+                    z_D = torch.randn_like(model_input) * sigma_data
+                    noised_predicted_x0 = torch.cos(t_D) * pred_x_0 + torch.sin(t_D) * z_D
+
+                    # Calculate adversarial loss
+                    pred_fake = disc(
+                        hidden_states=(noised_predicted_x0 / sigma_data),
+                        timestep=t_D.flatten(),
+                        encoder_hidden_states=prompt_embeds,
+                        encoder_attention_mask=prompt_attention_mask,
+                    )
+                    adv_loss = -torch.mean(pred_fake)
+
+                    # Total loss = sCM loss + LADD loss
+
+                    total_loss = args.scm_lambda * loss + adv_loss * args.adv_lambda
+
+                    total_loss = total_loss / args.gradient_accumulation_steps
+
+                    accelerator.backward(total_loss)
+
+                    g_step += 1
+
+                    if accelerator.sync_gradients:
+                        grad_norm = accelerator.clip_grad_norm_(transformer.parameters(), args.gradient_clip)
+                        if torch.logical_or(grad_norm.isnan(), grad_norm.isinf()):
+                            optimizer_G.zero_grad(set_to_none=True)
+                            optimizer_D.zero_grad(set_to_none=True)
+                            logger.warning("NaN or Inf detected in grad_norm, skipping iteration...")
+                            continue
+
+                        # switch phase to D
+                        phase = "D"
+
+                        optimizer_G.step()
+                        lr_scheduler.step()
+                        optimizer_G.zero_grad(set_to_none=True)
+
+            elif phase == "D":
```

Comment:
So this alternates between two phases in the _same_ training step, right? If so, I would add a comment. 

Also, should we let the users control the step interval in which the discriminator should be updated? Or not really?